docs/book/src/quinn/certificate.md

```diff
@@ -0,0 +1,163 @@
+
+In this section we look at certifying an endpoint with a certificate. 
+This can be done with either a non-self-signed or a self-signed certificate.
+
+### Self Signed Certificates
+
+Relying on [self-signed][5] certificates means that clients allow servers to sign their own certificates. 
+This is simpler because no third party has to be involved in signing the server's certificate.
+However, they do not offer all the security features that certificates from trusted CAs do have. 
+Self-signed certificates, among other options, can be created using the [rcgen][4] crate or the openssl binary.
+In this example [rcgen][4] is used.   
+
+Let's look at an example:
+
+```rust
+pub fn generate_self_signed_cert(cert_path: &str, key_path: &str) -> anyhow::Result<(quinn::Certificate, quinn::PrivateKey)> {
+    // Generate dummy certificate.
+    let certificate = rcgen::generate_simple_self_signed(vec!["localhost".into()]).unwrap();
+    let serialized_key = certificate.serialize_private_key_der();
+    let serialized_certificate = certificate.serialize_der().unwrap();
+
+    // Write to files.
+    fs::write(&cert_path, &serialized_certificate).context("failed to write certificate")?;
+    fs::write(&key_path, &serialized_key).context("failed to write private key")?;
+
+    let cert = quinn::Certificate::from_der(&cert)?;
+    let key = quinn::PrivateKey::from_der(&serialized_key)?;
+    Ok((cert, key))
+}
+```
+
+*Note that [generate_simple_self_signed][generate_simple_self_signed] returns a [Certificate][2] that can be serialized to both `.der` and `.pem` formats.*
+
+### Non-self-signed Certificates
+
+[Let's Encrypt][6] is a well-known Certificate Authority ([CA][1]) (certificate issuer) and distributes certificates for free.
+For this section lets-encrypt is used however any CA could be used interchangeably. 
+
+**Generate Certificate**
+
+Let's encrypt uses [Certbot][7] to generate certificates. 
+The certbot websites give clear instructions on how to use the tool.  
+Normally a certificate is generated to secure a web server, however, because we generate a certificate for a protocol,
+the configuration process will be slightly different than normal.
+If you do want to use an existing web server to generate certificates, please follow the instructions on certbot's website.
+
+Select on the certbot website that you do not have a web server (we assume you don't) and follow the given installation instructions.
+Certbot must answer a cryptographic challenge of the Let's Encrypt API to prove that you control the domain. 
+It uses ports 80 (HTTP) or 443 (HTTPS) to achieve this. Open the appropriate port in your firewall and router.
+
+If certbot is installed, run `certbot certonly --standalone`, this command will start a web server in the background and start the challenge.
+Certbot asks for the required data and writes the certificate to `cert.pem` and the private key to `privkey.pem`.  
+These files can then be referenced in code.  
```

Comment:
```suggestion
[certbot][7] can be used with Let's Encrypt to generate certificates; its website comes with clear instructions.
Because we're generating a certificate for an internal test server, the process used will be slightly different compared to what you would do when generating certificates for an existing (public) website.

On the certbot website, select that you do not have a public web server and follow the given installation instructions.
certbot must answer a cryptographic challenge of the Let's Encrypt API to prove that you control the domain. 
It needs to listen on port 80 (HTTP) or 443 (HTTPS) to achieve this. Open the appropriate port in your firewall and router.

If certbot is installed, run `certbot certonly --standalone`, this command will start a web server in the background and start the challenge.
certbot asks for the required data and writes the certificate to `cert.pem` and the private key to `privkey.pem`.  
These files can then be referenced in code.  
```